File: src/strategies/trie.js
import { normalizePathname, splitSegments } from "../normalize.js";
import { parsePattern } from "../pattern.js";

/**
 * @typedef {Object} TrieNode
 * @property {Map<string, TrieNode>} staticChildren - Child nodes indexed by literal segment.
 * @property {{name: string, node: TrieNode} | null} paramChild - Optional single parameter child for this level.
 * @property {{route: string, value: any} | null} terminal - The route stored at this node when a pattern ends here.
 * @property {{route: string, value: any} | null} wildcard - The optional "*" fallback route, only set on the root node.
 */

/**
 * Create an empty trie node.
 *
 * @returns {TrieNode} A fresh, unpopulated trie node.
 */
function createNode() {
  return {
    staticChildren: new Map(),
    paramChild: null,
    terminal: null,
    wildcard: null,
  };
}

/**
 * Insert a route entry into the trie.
 *
 * Pattern segments are walked one level per segment. Static segments descend
 * into staticChildren; ":name" segments descend into paramChild. When the
 * segments are exhausted, the route is stored as a terminal on the final node.
 *
 * @param {TrieNode} root - The trie root node.
 * @param {import("../pattern.js").PatternSegment[]} segments - Parsed pattern segments.
 * @param {{route: string, value: any}} entry - The route pattern and its opaque value.
 * @returns {void}
 */
function insert(root, segments, entry) {
  let node = root;
  for (const segment of segments) {
    if (segment.type === "param") {
      if (!node.paramChild) {
        node.paramChild = { name: segment.name, node: createNode() };
      }
      node = node.paramChild.node;
      continue;
    }
    let child = node.staticChildren.get(segment.value);
    if (!child) {
      child = createNode();
      node.staticChildren.set(segment.value, child);
    }
    node = child;
  }
  node.terminal = entry;
}

/**
 * Build a path trie from a compiled route collection.
 *
 * Static and dynamic routes share the same trie. The wildcard route, when
 * present, is attached to the root node as a global fallback.
 *
 * @param {import("../compile.js").CompiledRoutes} compiled - The compiled route collection.
 * @returns {TrieNode} The root trie node.
 */
function buildTrie(compiled) {
  const root = createNode();
  if (compiled.wildcard) {
    root.wildcard = compiled.wildcard;
  }

  for (const [raw, entry] of compiled.staticRoutes) {
    insert(root, parsePattern(raw).segments, entry);
  }
  for (const route of compiled.dynamicRoutes) {
    insert(root, route.segments, { route: route.route, value: route.value });
  }

  return root;
}

/**
 * Walk the trie for a list of segments and return the first match.
 *
 * Static children are preferred over the parameter child, so literal routes
 * always win over dynamic routes at the same level. The wildcard fallback is
 * not consulted here; it is applied by the caller when no terminal is found.
 *
 * @param {TrieNode} root - The trie root node.
 * @param {string[]} segments - The normalized pathname segments.
 * @returns {{terminal: {route: string, value: any}, params: Object<string, string>} | null} The match, or null when none.
 */
function matchTrie(root, segments) {
  let node = root;
  const params = {};

  for (let i = 0; i < segments.length; i++) {
    const segment = segments[i];
    const child = node.staticChildren.get(segment);
    if (child) {
      node = child;
      continue;
    }
    if (node.paramChild) {
      params[node.paramChild.name] = decodeURIComponent(segment);
      node = node.paramChild.node;
      continue;
    }
    return null;
  }

  if (node.terminal) {
    return { terminal: node.terminal, params };
  }
  return null;
}

/**
 * Create a matcher strategy backed by a path trie.
 *
 * Suitable for collections with a large number of routes, where a single trie
 * traversal scales better than scanning regular expressions. Static and
 * dynamic routes live in the same trie; an optional wildcard is the fallback.
 *
 * @param {import("../compile.js").CompiledRoutes} compiled - The compiled route collection.
 * @returns {import("../compile.js").MatchStrategy} A strategy exposing a match method.
 */
export function createTrieStrategy(compiled) {
  const root = buildTrie(compiled);

  return {
    match(pathname) {
      const normalized = normalizePathname(pathname);
      const segments = splitSegments(normalized);

      const matched = matchTrie(root, segments);
      if (matched) {
        return {
          found: true,
          pathname: normalized,
          route: matched.terminal.route,
          params: matched.params,
          value: matched.terminal.value,
        };
      }

      if (root.wildcard) {
        return {
          found: true,
          pathname: normalized,
          route: root.wildcard.route,
          params: {},
          value: root.wildcard.value,
        };
      }

      return {
        found: false,
        pathname: normalized,
        route: null,
        params: {},
        value: null,
      };
    },
  };
}

File: python/roution/strategies/trie.py
from urllib.parse import unquote

from roution.compile import StaticRoute
from roution.normalize import normalize_pathname, split_segments, parse_query
from roution.pattern import parse_pattern
from roution.result import make_result


class _TrieNode:
    __slots__ = ("static_children", "param_child", "terminal", "wildcard")

    def __init__(self):
        self.static_children = {}
        self.param_child = None  # (name, _TrieNode)
        self.terminal = None  # StaticRoute
        self.wildcard = None  # StaticRoute


def _insert(root, segments, entry):
    node = root
    for segment in segments:
        if segment.type == "param":
            if node.param_child is None:
                node.param_child = (segment.name, _TrieNode())
            node = node.param_child[1]
            continue
        child = node.static_children.get(segment.value)
        if child is None:
            child = _TrieNode()
            node.static_children[segment.value] = child
        node = child
    node.terminal = entry


def _build_trie(compiled):
    root = _TrieNode()
    if compiled.wildcard is not None:
        root.wildcard = compiled.wildcard
    for raw, static_route in compiled.static_routes.items():
        _insert(root, parse_pattern(raw).segments, static_route)
    for dynamic_route in compiled.dynamic_routes:
        _insert(
            root,
            dynamic_route.segments,
            StaticRoute(route=dynamic_route.route, value=dynamic_route.value),
        )
    return root


def _match_trie(root, segments):
    node = root
    params = {}
    for segment in segments:
        child = node.static_children.get(segment)
        if child is not None:
            node = child
            continue
        if node.param_child is not None:
            name, child = node.param_child
            params[name] = unquote(segment)
            node = child
            continue
        return None, None
    if node.terminal is not None:
        return node.terminal, params
    return None, None


class TrieStrategy:
    """Matcher strategy backed by a path trie.

    Static and dynamic routes live in the same trie; an optional wildcard is
    the fallback. Suitable for collections with a large number of routes.
    """

    def __init__(self, compiled, include_query):
        self.compiled = compiled
        self.include_query = include_query
        self.root = _build_trie(compiled)

    def match(self, pathname):
        normalized = normalize_pathname(pathname)
        query = parse_query(pathname) if self.include_query else {}
        segments = split_segments(normalized)

        terminal, params = _match_trie(self.root, segments)
        if terminal is not None:
            return make_result(
                True, normalized, terminal.route, params, query, terminal.value
            )

        if self.compiled.wildcard is not None:
            wildcard = self.compiled.wildcard
            return make_result(
                True, normalized, wildcard.route, {}, query, wildcard.value
            )

        return make_result(False, normalized, None, {}, query, None)


def create_trie_strategy(compiled, include_query):
    return TrieStrategy(compiled, include_query)
